### DISCUSSION TITLE
Weights Generation

### DISCUSSION THREAD

**Discussion by erex16** [bcollazo/catanatron#256] [Q&A] [answered]
Hello! Thanks for sharing this project, you've clearly put a lot of work into it and it's super cool. I'm interested in creating my own bot and potentially messing around with some alternative ML techniques. In looking over your value.py file that shows all the weights for various action possibilities, I'm curious how you generated these numbers? Was it all trial and error? The contender weights seem especially specific and I was wondering if you have some algorithm or model that you used to generate these. Knowing how these were determined would be helpful as I try to create new weights.

**Comment by bcollazo** [comment, marked as answer]
Hello! Great to hear!

It was completely trial and error / guessing / my own intuition on what I think might be good. From there you are right that the Contender Weights are particularly specific, I don't remember exactly how I got to those, but it must have been from either:
- Applying the "SPSA" algorithm talked about in https://www.chessprogramming.org/SPSA and implemented here: https://github.com/bcollazo/catanatron/blob/master/catanatron_experimental/catanatron_experimental/spsa.py
- Trying "Bayesian Optimization" (see https://github.com/bayesian-optimization/BayesianOptimization)
- Or using "optuna" (see https://github.com/bcollazo/catanatron/blob/master/catanatron_experimental/catanatron_experimental/optunation.py)

I haven't tried out "Genetic Algorithms" but that might be another way to improve / locally optimize the hand-crafted weights further.

I think one of the main problems the bot has right now is that the value function seems to optimize for local value (it can't see many turns ahead / think longterm). This is particularly seen with the road-building features. I think there might be a better way to capture how good a player's road structure is, than what is in the repo now.

Hope you can find an improvement!
